sdk/keyvault/keyvault/src/secrets/index.ts

```diff
@@ -0,0 +1,470 @@
+
+import {
+  HttpClient as IHttpClient,
+  HttpPipelineLogger as IHttpPipelineLogger,
+  ServiceClientCredentials,
+  RequestPolicyFactory,
+  deserializationPolicy,
+  signingPolicy,
+  RequestOptionsBase,
+  exponentialRetryPolicy,
+  redirectPolicy,
+  systemErrorRetryPolicy,
+  generateClientRequestIdPolicy,
+  proxyPolicy,
+  throttlingRetryPolicy,
+  getDefaultProxySettings,
+  userAgentPolicy
+} from "@azure/ms-rest-js";
+
+import { AzureServiceClientOptions as Pipeline, getDefaultUserAgentValue } from "@azure/ms-rest-azure-js";
+import { RetryOptions, ProxyOptions, TelemetryOptions } from "..";
+import { SecretBundle } from "../models";
+import { KeyVaultClient } from "../keyVaultClient";
+import { RetryConstants, SDK_VERSION } from "../utils/constants";
+import {
+  Secret,
+  DeletedSecret,
+  SetSecretOptions,
+  UpdateSecretOptions,
+  GetSecretOptions,
+  GetAllSecretsOptions,
+  SecretAttributes
+} from "./secretsModels";
+import { parseKeyvaultIdentifier as parseKeyvaultEntityIdentifier } from "../utils";
+
+export { ParsedKeyVaultEntityIdentifier } from "./secretsModels";
+export { Pipeline };
+
+/**
+ * Option interface for Pipeline.newPipeline method.
+ *
+ * Properties of this interface should not overlap with properties of {@link Pipeline}
+ * as we use them to differentiate instances of NewPipelineOptions from instances of Pipeline.
+ * If this interface is modified, the method isNewPipelineOptions() should also be updated
+ * to adapt the changes.
+ *
+ * @export
+ * @interface NewPipelineOptions
+ */
+export interface NewPipelineOptions {
+  /**
+   * Telemetry configures the built-in telemetry policy behavior.
+   *
+   * @type {TelemetryOptions}
+   * @memberof NewPipelineOptions
+   */
+  telemetry?: TelemetryOptions;
+  retryOptions?: RetryOptions;
+  proxyOptions?: ProxyOptions;
+
+  logger?: IHttpPipelineLogger;
+  HTTPClient?: IHttpClient;
+}
+
+function isNewPipelineOptions(
+  pipelineOrOptions: Pipeline | NewPipelineOptions
+): pipelineOrOptions is NewPipelineOptions {
+  // An empty object is consider options
+  function isEmptyObject(obj: Pipeline | NewPipelineOptions) {
+    return Object.keys(obj).length === 0 && obj.constructor === Object;
+  }
+  const options = pipelineOrOptions as NewPipelineOptions;
+  return (
+    isEmptyObject(pipelineOrOptions) ||
+    !!(options.retryOptions || options.proxyOptions || options.logger || options.HTTPClient)
+  );
+}
+
+export class SecretsClient {
+  /**
+   * A static method used to create a new Pipeline object with the provided Credential.
+   *
+   * @static
+   * @param {ServiceClientCredentials} credential that implements signRequet().
+   * @param {NewPipelineOptions} [pipelineOptions] Optional. Options.
+   * @returns {Pipeline} A new Pipeline object.
+   * @memberof SecretsClient
+   */
+  public static getDefaultPipeline(
+    credential: ServiceClientCredentials,
```

Comment:
>credential: ServiceClientCredentials, [](start = 4, length = 37)

For now ServiceClientCredentials will do.  Once we have the Auth updates for the js repo this will change to TokenCredential from @azure/core

Reply:
@schaabs - I just opened up one more possibility to authenticate to Keyvault via azure cli. Some customers love it. Hence keeping the interface to ServiceClientCredentials would be very useful.

- [Link](https://github.com/Azure/azure-sdk-for-node/issues/2284#issuecomment-493786455) to the customer issue
- Sample solution
```typescript
import { AzureCliCredentials } from "@azure/ms-rest-nodeauth";
import { KeyVaultClient, KeyVaultModels, KeyVaultMappers } from "@azure/keyvault";

async function main(): Promise<void> {
  try {
    const creds = AzureCliCredentials.create({ resource: "https://vault.azure.net" });
    const client = new KeyVaultClient(creds);
    const vaultBaseUrl = 'testvaultBaseUrl';
    const keyName = 'testkeyName';
    const keyVersion = 'testkeyVersion';
    const result = await client.getKey(vaultBaseUrl, keyName, keyVersion);
    console.log(result);
  } catch (err) {
    console.log(err);
  }
}

main();
```

- **This makes logging super simple for node.js scenario.** 
- The credential takes care of refreshing the token if the token expires OR the subscription was changed OR the resource (token audience) was changed.

Reply:
One other thing that we did in the old node sdk and customers loved that too. Customers hated the authenticator challenge callback flip flop. Their ask was, if they have already authenticated to Azure via one of the login methods available in ms-rest-azure such as `interactiveLogin()`, `loginWithServicePrincipalSecret`, `loginWithVmMSI`, etc. then why should they authenticate again in their App?

Thus we provided them a much simpler interface where, the sdk was internally converting the management plane credentials into KeyVaultCredentials. It works like a charm and customers love it.

Take a look at that sample [here](https://github.com/Azure/azure-sdk-for-node/tree/master/lib/services/keyVault#authentication-and-client-creation).

IMHO, rather than creating a new construct for credentials for node.js scenario, it would be nice to stick to existing ServiceClientCredentials interface and may be extend that interface rather than introducing a new interface and make customers authenticate twice.

Reply:
I created this issue to address `Use TokenCredential in the Secrets`: https://github.com/Azure/azure-sdk-for-js/issues/3029

Two questions:
- Should we create one for `[JS] [KeyVault] Authentication method: azure -cli` ?
- Should we create an issue for `[JS] [KeyVault] Authentication method: converting the management plane credentials into KeyVaultCredentials`?

Reply:
- Authentication via `azure-cli` will only work for node.js scenario and that has already been implemented in the new TS based auth library for nodejs [@azure/ms-rest-nodeauth](https://github.com/Azure/ms-rest-nodeauth/blob/master/lib/credentials/azureCliCredentials.ts#L321).

- Converting management plane credentials into KeyVaultCredentials needs to happen and creating a work item for the same will be helpful. Part of that work has already been done in `@azure/ms-rest-nodeauth` over [here](https://github.com/Azure/ms-rest-nodeauth/blob/master/lib/credentials/keyVaultFactory.ts).
   - The remaining part is to create a signing policy in the request pipeline that will get the initial challenge and send the request to get the access token accordingly. You can take a look over [here](https://github.com/Azure/azure-sdk-for-node/blob/master/runtime/ms-rest-azure/lib/credentials/keyVaultCredentials.js#L108) for the full implementation in the old node sdk.